src/library_glfw.js

```diff
@@ -671,6 +672,64 @@ var LibraryGLFW = {
       win.scrollFunc = cbfun;
     },
 
+    setDropCallback: function(winid, cbfun) {
+      var win = GLFW.WindowFromId(winid);
+      if (!win) return;
+      win.dropFunc = cbfun;
+    },
+
+    onDrop: function(event) {
+      if (!GLFW.active.dropFunc) return;
+
+      event.preventDefault();
+
+      var filenames = allocate(new Array(event.dataTransfer.files.length*4), 'i8*', ALLOC_NORMAL);
+      var filenamesArray = [];
+      var count = event.dataTransfer.files.length;
+
+      // Read and save the files to emscripten's FS
+      // TODO: lazily load?
+      var written = 0;
+      for (var i = 0; i < count; ++i) {
+        (function(file) {
+          var reader = new FileReader();
+          reader.onload = function(e) {
+            var data = e.target.result;
+            var path = file.name; // TODO: to a new directory?
+            FS.writeFile(path, new Uint8Array(data), { encoding: 'binary' });
+            //console.log('wrote '+path+', size '+data.byteLength+', count '+written+' of '+count);
+            if (++written === count) {
+              console.log('calling callback');
+              Module['dynCall_viii'](GLFW.active.dropFunc, GLFW.active.id, count, filenames);
+
+              for (var i = 0; i < filenamesArray.length; ++i) _free(filenamesArray[i]);
+              _free(filenames);
+            }
+          };
+          reader.readAsArrayBuffer(file);
+
+          var filename = allocate(intArrayFromString(file.name), 'i8', ALLOC_NORMAL);
+          filenamesArray.push(filename);
+          setValue(filenames + i*4, filename, 'i8*');
+        })(event.dataTransfer.files[i]);
```

Comment:
The [IIFE](https://en.wikipedia.org/wiki/Immediately-invoked_function_expression) is to pass the correct file in each iteration of the loop to the reader.onload closure, without it `var file` would lexically capture only the last file.

The usual way to solve this is use `.forEach(function() { ... })` but unfortunately `event.dataTransfer.files` is not an `Array` but a `FileList` so it doesn't have `.forEach`, leading to this awkward construction.

I'll see if I can refactor it to avoid the IIFE...